apps/cyberstorm-nextjs/utils/auth.ts

```diff
@@ -0,0 +1,129 @@
+import { NextResponse } from "next/server";
+import type { NextRequest } from "next/server";
+import hkdf from "@panva/hkdf";
+import { EncryptJWT, jwtDecrypt } from "jose";
+import { v4 as uuid } from "uuid";
+
+interface UserStatusType {
+  username: string | null;
+  capabilities: string[];
+  subscription: { expires: string | null };
+  rated_packages: string[];
+  teams: string[];
+}
+
+const now = () => (Date.now() / 1000) | 0;
+
+async function getDerivedEncryptionKey(secret: string | Buffer) {
+  return await hkdf(
+    "sha256",
+    secret,
+    "",
+    "Cyberstorm Generated Encryption Key",
+    32
+  );
+}
+
+export async function encode(token: { username: string; expires: number }) {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    return await new EncryptJWT(token)
+      .setProtectedHeader({ alg: "dir", enc: "A256GCM" })
+      .setIssuedAt()
+      .setExpirationTime(Math.min(now() + 900, token.expires))
+      .setJti(uuid())
+      .encrypt(encryptionSecret);
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+export async function decode(
+  token: string
+): Promise<Record<string, unknown> | null> {
+  if (process.env.CYBERSTORM_AUTH_SECRET) {
+    if (!token) return null;
+    const encryptionSecret = await getDerivedEncryptionKey(
+      process.env.CYBERSTORM_AUTH_SECRET
+    );
+    const { payload } = await jwtDecrypt(token, encryptionSecret, {
+      clockTolerance: 15,
+    });
+    return payload;
+  } else {
+    throw new Error("CYBERSTORM_AUTH_SECRET is missing");
+  }
+}
+
+async function fetchSubStatus(sessionid: string | undefined): Promise<{
+  username: string | null;
+  subActive: boolean;
+  expires: number;
+}> {
+  if (sessionid) {
+    const res = await fetch(
+      `https://thunderstore.io/api/experimental/current-user/`,
+      {
+        headers: {
+          authorization: `Session ${sessionid}`,
+          "Content-Type": "application/json",
+        },
+      }
+    );
+    if (res.status === 200) {
+      const currentUserStatus: UserStatusType = await res.json();
+      if (currentUserStatus.subscription.expires) {
+        const expiryTimeEpoch = Date.parse(
+          currentUserStatus.subscription.expires
+        );
+        if (expiryTimeEpoch < now()) {
+          return {
+            username: currentUserStatus.username,
+            subActive: true,
+            expires: expiryTimeEpoch,
+          };
+        }
+      }
+    }
+  }
+  return { username: null, subActive: false, expires: 0 };
+}
+
+export function redirectToPremium(
+  request: NextRequest,
+  response: NextResponse
+) {
+  if (request.nextUrl.pathname !== "/premium") {
```

Comment:
I'm a bit confused due context swapping to NextRequest and NextResponse. Why do we check the request path here? Does it contain the original path the user actually requested? Can some redirect already have changed this? Would it be better to just redirect user to the premium page no matter what?

Reply:
The request contains the path, to the user requested page and if we don't want to redirect endlessly we'll need to make sure the redirection is not done on the `/premium` page. Although, now I'm, not entirely sure if next could handle this fine after all.

I'll check and test this out, when I get the local env fully setup.

Reply:
Checked this out, worked nicely with this and without it just kept redirecting endlessly